packages/core/src/registry/StyleSheetRegistry.ts

```diff
@@ -7,93 +8,105 @@ const STYLE_ID_PREFIX = "__";
+      compile(css).forEach((element) => {
+        const rule = serialize([element], stringify);
+        this.indexesMap[id] = (this.indexesMap[id] || []).concat(
+          this.sheet.insertRule(rule)
+        );
+      });
     }
   }
 
   public remove(id: string): void {
-    if (
-      this.indices[id] === undefined ||
-      this.instancesCounts[id] === undefined
-    ) {
-      throw new Error(`StyleSheetRegistry: id: \`${id}\` not found.`);
+    if (this.instancesCountMap[id] === undefined) {
+      throw new Error(
+        `StyleSheetRegistry: id: \`${id}\` not found in idInstancesCountMap.`
+      );
     }
 
     // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    this.instancesCounts[id]! -= 1;
+    this.instancesCountMap[id]! -= 1;
     // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    if (this.instancesCounts[id]! !== 0) {
+    if (this.instancesCountMap[id]! !== 0) {
       return;
     }
 
-    const serverSideRenderedStyle = this.serverSideRenderedStyles?.[id];
+    const serverSideRenderedStyle = this.serverSideRenderedStyleMap?.[id];
     if (serverSideRenderedStyle) {
       serverSideRenderedStyle.remove();
-      delete this.serverSideRenderedStyles?.[id];
+      delete this.serverSideRenderedStyleMap?.[id];
     } else {
+      if (this.indexesMap[id] === undefined) {
+        throw new Error(
+          `StyleSheetRegistry: id: \`${id}\` not found in idIndexesMap.`
+        );
+      }
       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-      this.sheet.deleteRule(this.indices[id]!);
+      this.indexesMap[id]!.forEach((index) => this.sheet.deleteRule(index));
+      delete this.indexesMap[id];
     }
 
-    delete this.indices[id];
-    delete this.instancesCounts[id];
+    delete this.instancesCountMap[id];
   }
 
   public styles(options: { nonce?: string } = {}) {
-    return Object.keys(this.indices)
+    return Object.keys(this.indexesMap)
       .map((id) => {
-        const index = this.indices[id];
-        if (index === undefined) {
+        const indexes = this.indexesMap[id];
+        if (indexes === undefined) {
           return null;
         }
         const cssRules = this.sheet.cssRules();
-        const rule = cssRules[index];
-        if (rule === undefined) {
+        const css = indexes
+          .map((index) => cssRules[index]?.cssText)
+          .filter(Boolean)
+          .join(this.sheet.isSpeedy() ? "" : "\n");
+        if (css.length === 0) {
           return null;
         }
         return React.createElement("style", {
           id: `${STYLE_ID_PREFIX}${id}`,
           key: `${STYLE_ID_PREFIX}${id}`,
           nonce: options.nonce ? options.nonce : undefined,
           dangerouslySetInnerHTML: {
-            __html: rule.cssText,
+            __html: css,
           },
         });
       })
-      .filter((props) => props !== null);
+      .filter(Boolean);
```

Comment:
added code to concatenate the CSS rules obtained from the indexes because fixed to have multiple indexes per id (className).